### PR TITLE
1 create initial processing chain

### PR DESCRIPTION
**Pull request by rosepearson**
A currently draft pull request where we can discuss the form of the initial tests associated with the package.

The package 'GeoFabrics' loads in data specified in an instruction.json file and runs a bare-bones processing chain to produce a water-tight DEM over the specified catchment. 

_Later issues will deal with ensuring that the produced DEM is hydrologically conditioned._

The ./src/main.py and ./instruction.json files currently act as examples for usage. These will be replaced by the test1 under root /tests/test1

_**A still unfulfilled requirement of the PR in my mind is replacement/utilisation of these by the ./tests/test1 initial test.**_ As part of this PR I would like to get input on the form of this test before creating synthetic data.

### CODE REVIEW

File: src/GeoFabrics/geometry.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        """ Return the catchment foreshore region """
+        
+        if self._foreshore is None:
+            self._foreshore = geopandas.overlay(self.land_and_foreshore, 
+                                                self.land, how='difference')
+        return self._foreshore
+    
+    @property    
+    def land_and_foreshore(self):
+        """ Return the catchment and and foreshore region """
```
Comment (LukeParky) [inline src/GeoFabrics/geometry.py:100]: catchment and _land_ foreshore region

File: src/GeoFabrics/geometry.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    
+    @property
+    def lidar_extents(self):
+        """ Return the catchment lidar extents """
+        
+        assert self._lidar_extents is not None, "lidar_extents have not been set, and need to be set explicitly"
+            
+        return self._lidar_extents
+    
+    def load_lidar_extents(self, lidar_extents_string): # expect to be filepath
```
Comment (LukeParky) [inline src/GeoFabrics/geometry.py:133]: You could do this type comment as a type hint such as in [type hint](https://docs.python.org/3/library/typing.html)  or other ways with in-built Python support

File: instruction.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
+{"instructions": 
+  {
+    "projection": 2193,
+	"grid_params": {
+      "resolution": 10
+    },
+    "data_paths": {
+      "catchment_boundary": "C:\\Users\\pearsonra\\Documents\\data\\Wakanae\\Small_test_site\\initial_data\\boundary\\boundary.shp",
```
Comment (LukeParky) [inline instruction.json:8]: Could these be in a relative path format, such as in the test example?

**Review by LukeParky — CHANGES_REQUESTED** [review]
Looks like good structure, just some minor changes that I suggested

File: src/GeoFabrics/dem.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    """
+    
+    def __init__(self, dem_file, catchment_geometry, set_foreshore = True):
+        """ Load in dem """
+        
+        self.catchment_geometry = catchment_geometry
+        self.set_foreshore = set_foreshore
+        self._dem = rioxarray.rioxarray.open_rasterio(dem_file, masked=True)
+        
+        self.__set_up()
```
Comment (LukeParky) [inline src/GeoFabrics/dem.py:25]: I'm unsure of the use of a double underscore here, compared to a single one. I'm not telling you to change it but just highlighting it

File: instruction.json (hunk 3 above)
Comment (rosepearson) [reply instruction.json:8]: They could be - I was thinking of removing this from version control once the test is created and working. I will leave for now and remove prior to the end of this pr.

**Comment by rosepearson** [comment]
I'm planning to make a test dataset with the topography represented in the attached image. 

The datasets would be:
* LiDAR - 1 pt per metre over the middle 500x500m2 area as a LAZ file (starting 25m offshore)
* Reference DEM - a 15m resolution extending over the entire land region and extending 25m offshore
* Bathymetry contours - a 5m, 10m and 20m depth contour
 As well as the land and catchment bounding boxes in the image.

@AliceHarang, what do you think?

image

File: src/GeoFabrics/dem.py (hunk 4 above)
Comment (rosepearson) [reply src/GeoFabrics/dem.py:25]: I did some reading (link to docs below), which indicates that a double underscore should be used if you want to do name-mangling and a single underscore should be used for private variables - which is what I'm after so I'll change it thanks! 

https://docs.python.org/3/tutorial/classes.html#private-variables

**Comment by AliceHarang** [comment]
The topographic test dataset seems great ! Perhaps you can take a smaller region at the back behind the "top-range" line...
Do you think about adding some perturbations (noise or larger ones) in it?

File: src/GeoFabrics/geometry.py (hunk 2 above)
Comment (rosepearson) [reply src/GeoFabrics/geometry.py:133]: Good idea. I've added for most of the class constructors as well

**Comment by rosepearson** [comment]
> The topographic test dataset seems great ! Perhaps you can take a smaller region at the back behind the "top-range" line...
> Do you think about adding some perturbations (noise or larger ones) in it?

Yes - I wondered that too. Maybe I'll just have the DEM extend just beyond the range-top.

I was also wondering about adding some noise - but thought I wouldn't for the very first test case (keep it very simple) and that could be done later as a whole new test. I could probably do it for this PR though.

**Review by LukeParky — CHANGES_REQUESTED** [review]
Looks good to me. Resolve the conversation on `instruction.json` and it's approved

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    extras_require={
+        'dev': ['check-manifest'],
+        # 'test': ['coverage'],
+    },
+    # entry_points={
+    #     'console_scripts': [  # This can provide executable scripts
+    #         'run=examplepy:main',
+    # You can execute `run` in bash to run `main()` in src/examplepy/__init__.py
+    #     ],
+    # },
```
Comment (LukeParky) [inline setup.py:53]: Remove commented code

File: src/GeoFabrics/processor.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        
+        ### load in dense DEM 
+        metadata=json.loads(pdal_pipeline.get_metadata())
+        self.dense_dem = dem.DenseDem(metadata['metadata']['writers.gdal']['filename'][0], self.catchment_geometry)
+        
+        ### Load in bathy
+        z_label = self.instructions['instructions']['instructions']['bathymetry_contours_z_label'] if  \
+            'bathymetry_contours_z_label' in self.instructions['instructions']['instructions'] else None
+        self.bathy_contours = geometry.BathymetryContours(self.instructions['instructions']['data_paths']['bathymetry_contours'][0], self.catchment_geometry, z_label = z_label)
+        #bathy_points = geometry.BathymetryPoints(self.instructions['instructions']['data_paths']['bathymetry_points'][0], catchment_geometry)
```
Comment (LukeParky) [inline src/GeoFabrics/processor.py:99]: remove commented code

**Review by LukeParky — CHANGES_REQUESTED** [review]

File: setup.py (hunk 5 above)
Comment (rosepearson) [reply setup.py:53]: done!

**Review by LukeParky — CHANGES_REQUESTED** [review]
Unit test isn't passing on my machine

[test_output.txt](https://github.com/rosepearson/GeoFabrics/files/6744125/test_output.txt)

File: README.md [hunk 7]
```diff
@@ -1,18 +1,23 @@
...
 Basic code stub looks like:
-```
+```python
 import GeoFabrics
 import json
-with open(r'path\to\file\', 'r') as file_pointer:
+with open(r'path\to\file\\', 'r') as file_pointer:
             instructions = json.load(file_pointer)
-runner = processor.GeoFabricsGenerator(instructions)
+runner = GeoFabrics.processor.GeoFabricsGenerator(instructions)
```
Comment (LukeParky) [inline README.md]: geofabrics, not GeoFabrics

**Review by LukeParky — CHANGES_REQUESTED** [review]

**Comment by rosepearson** [comment]
Hi Luke,
I think I have addressed all of your suggestions, and the package is now ready to be tested.
Specifically, the testing I did is: 
1. run main.py with my own data and instruction file (I don't think you need to do this), 
2. run the unit tests and check they pass. 
3. import the package using pip install (see readme) into a different conda environment and check you can import geofabrics in Python.

File: README.md (hunk 7 above)
Comment (rosepearson) [reply README.md]: thanks - fixed

**Comment by rosepearson** [comment]
> Unit test isn't passing on my machine
> 
> [test_output.txt](https://github.com/rosepearson/GeoFabrics/files/6744125/test_output.txt)

Thanks! I've moved to making an almost equal comparison as discussed as the differences is in the order e-13 and only present in around 20% of the locations.

**Review by LukeParky — APPROVED** [review]

**Pull request merged**